core/shared/src/main/scala-3.x/monocle/internal/focus/features/castas/CastAsParser.scala

```diff
@@ -0,0 +1,22 @@
+package monocle.internal.focus.features.castas
+
+import monocle.internal.focus.FocusBase
+
+private[focus] trait CastAsParser {
+  this: FocusBase => 
+
+  import macroContext.reflect._
+
+  object CastAs extends FocusParser {
+
+    def unapply(term: Term): Option[FocusResult[(Term, FocusAction)]] = term match {
+      case Apply(TypeApply(Ident("as"), List(typeArg)), List(remainingCode)) => 
```

Comment:
Should we require the current type to be a sealed trait / enum and `toType` to be a subtype of it?

Reply:
This should be the only safe use case and the more constrained we are, the better the error message we can produce. WDYT?

Reply:
It is a more general discussion about what one should or should not be able to do. I would argue that we should only be able to pattern match/down cast on a sealed trait/abstract class or final class.
Let's say it otherwise, in which case would want to use `as` and the source type is not sealed?

This may not be the right place to have this discussion. I think it is awesome what you did with `as`, but in case you would like to echange more on this topic, I am happy to do it. Maybe not in this PR :)

Reply:
> Also, it wouldn't be supported in the standalone `as`, so it wouldn't commute with Focus

Very good point. I guess we could do it if the standalone `as` was implemented via a macro (if that's possible), but even then, it wouldn't work for Scala 2.12. So you are right, let's not add any more constraints.